File: src/cmc/builtins.py
"""The helper words (builtin clumps) of CMC.

Every helper checks its arguments and complains with a friendly OOGA! when
something is the wrong kind of thing.
"""

import math
import random

from .errors import CmcRuntimeError
from .values import show, type_name, number, whole
from . import words

CONSTANTS = {
    "gronk": True,
    "nork": False,
    "plop": None,
    "pi": math.pi,
}

# Arity for every helper, pulled from the caveman dictionary.
ARITY = {}
for _w in words.BUILTIN_WORDS:
    _call_name = _w.get("display", _w["name"])
    ARITY[_call_name] = _w["arity"]


def _need_text(value, line, extra):
    if not isinstance(value, str):
        raise CmcRuntimeError(
            extra + " needs text (goop), but got " + type_word(value) + ".",
            line=line,
            hint='put quotes around it, like "hello".',
        )
    return value


def _need_pile(value, line, extra):
    if not isinstance(value, list):
        raise CmcRuntimeError(
            extra + " needs a pile (snorf), but got " + type_word(value) + ".",
            line=line,
            hint="make a pile like this: grunk xs = snorf(1, 2, 3)",
        )
    return value


def _spot(value, line, extra):
    """A pile spot number (whole, may be negative)."""
    idx = whole(value, line=line, extra=extra)
    return idx


def bi_goop(interp, args, line):
    return show(args[0])


def bi_snorf(interp, args, line):
    return list(args)


def bi_nom(interp, args, line):
    value = args[0]
    if isinstance(value, (str, list)):
        return len(value)
    raise CmcRuntimeError(
        "nom counts piles and text, but got " + type_word(value) + ".",
        line=line,
        hint="nom(pets) counts a pile. nom(\"cave\") counts letters.",
    )


def bi_skoop(interp, args, line):
    value, idx = args[0], _spot(args[1], line, "skoop")
    if not isinstance(value, (str, list)):
        raise CmcRuntimeError(
            "skoop reaches into piles and text, but got " + type_word(value) + ".",
            line=line,
            hint="make a pile like this: grunk xs = snorf(1, 2, 3)",
        )
    if idx < 0:
        idx += len(value)
    if idx < 0 or idx >= len(value):
        raise CmcRuntimeError(
            "There is no spot " + show(idx) + " in that " + ("text" if isinstance(value, str) else "pile")
            + " (it has " + str(len(value)) + " spots).",
            line=line,
            hint="spots start at 0, so the last spot is " + str(len(value) - 1) + ".",
        )
    return value[idx]


def bi_yoink(interp, args, line):
    pile = _need_pile(args[0], line, "yoink")
    if not pile:
        raise CmcRuntimeError(
            "The pile is empty, so there is nothing to yoink!",
            line=line,
            hint="check nom(pile) > 0 before yoinking.",
        )
    return pile.pop()


def bi_plop(interp, args, line):
    pile = _need_pile(args[0], line, "plop")
    pile.append(args[1])
    return None


def bi_shout(interp, args, line):
    return _need_text(args[0], line, "shout").upper()


def bi_whisper(interp, args, line):
    return _need_text(args[0], line, "whisper").lower()


def bi_flip(interp, args, line):
    return _need_text(args[0], line, "flip")[::-1]


def bi_find(interp, args, line):
    text = _need_text(args[0], line, "find")
    piece = _need_text(args[1], line, "find")
    return text.find(piece)


def bi_split(interp, args, line):
    text = _need_text(args[0], line, "split")
    sep = _need_text(args[1], line, "split")
    if sep == "":
        raise CmcRuntimeError(
            "split needs something to cut on, and an empty text cuts nothing.",
            line=line,
        )
    return text.split(sep)


def bi_join(interp, args, line):
    pile = _need_pile(args[0], line, "join")
    sep = show(args[1]) if not isinstance(args[1], str) else args[1]
    return sep.join(show(item) for item in pile)


def bi_what(interp, args, line):
    return type_name(args[0])


def bi_numba(interp, args, line):
    value = args[0]
    if isinstance(value, bool):
        raise CmcRuntimeError(
            "numba cannot turn gronk/nork into a number.",
            line=line,
            hint="use 1 for gronk or 0 for nork.",
        )
    if isinstance(value, (int, float)):
        return value
    if isinstance(value, str):
        text = value.strip()
        try:
            return int(text)
        except ValueError:
            pass
        try:
            return float(text)
        except ValueError:
            raise CmcRuntimeError(
                "numba cannot turn \"" + value + "\" into a number.",
                line=line,
                hint='answers that are numbers look like "42" or "3.5".',
            )
    raise CmcRuntimeError(
        "numba needs text or a number, but got " + type_word(value) + ".",
        line=line,
        hint='use it like this: numba("42")',
    )


def bi_munga(interp, args, line):
    lo = number(args[0], line=line, extra="munga")
    hi = number(args[1], line=line, extra="munga")
    if lo > hi:
        lo, hi = hi, lo
    if isinstance(lo, int) and isinstance(hi, int):
        return random.randint(lo, hi)
    return random.uniform(lo, hi)


def bi_round(interp, args, line):
    value = number(args[0], line=line, extra="round")
    return int(math.floor(value + 0.5)) if value >= 0 else -int(math.floor(-value + 0.5))


def bi_flat(interp, args, line):
    return math.floor(number(args[0], line=line, extra="flat"))


def bi_roof(interp, args, line):
    return math.ceil(number(args[0], line=line, extra="roof"))


def bi_abs(interp, args, line):
    return abs(number(args[0], line=line, extra="abs"))


def bi_small(interp, args, line):
    a = number(args[0], line=line, extra="small")
    b = number(args[1], line=line, extra="small")
    return a if a <= b else b


def bi_big(interp, args, line):
    a = number(args[0], line=line, extra="big")
    b = number(args[1], line=line, extra="big")
    return a if a >= b else b


def bi_root(interp, args, line):
    value = number(args[0], line=line, extra="root")
    if value < 0:
        raise CmcRuntimeError(
            "root cannot work on negative numbers (nothing times itself makes " + show(value) + ").",
            line=line,
            hint="use abs(value) first if you want the size without the minus.",
        )
    result = math.sqrt(value)
    return int(result) if result.is_integer() else result


def bi_pow(interp, args, line):
    base = number(args[0], line=line, extra="pow")
    times = number(args[1], line=line, extra="pow")
    result = base ** times
    if isinstance(result, float) and result.is_integer():
        return int(result)
    return result


def bi_wait(interp, args, line):
    ms = number(args[0], line=line, extra="wait")
    if ms < 0:
        ms = 0
    interp.wait_ms(ms)
    return None


IMPLS = {
    "goop": bi_goop,
    "snorf": bi_snorf,
    "nom": bi_nom,
    "skoop": bi_skoop,
    "yoink": bi_yoink,
    "plop": bi_plop,
    "shout": bi_shout,
    "whisper": bi_whisper,
    "flip": bi_flip,
    "find": bi_find,
    "split": bi_split,
    "join": bi_join,
    "what": bi_what,
    "numba": bi_numba,
    "munga": bi_munga,
    "round": bi_round,
    "flat": bi_flat,
    "roof": bi_roof,
    "abs": bi_abs,
    "small": bi_small,
    "big": bi_big,
    "root": bi_root,
    "pow": bi_pow,
    "wait": bi_wait,
}

HELPER_NAMES = set(IMPLS)


def call_builtin(interp, name, args, line):
    impl = IMPLS.get(name)
    if impl is None:
        return None, False
    low, high = ARITY[name]
    if len(args) < low or len(args) > high:
        if low == high:
            wanted = str(low)
        else:
            wanted = str(low) + " to " + str(high)
        raise CmcRuntimeError(
            "'" + name + "' wants " + wanted + " thing" + ("s" if wanted != "1" else "")
            + ", but got " + str(len(args)) + ".",
            line=line,
            hint="look at the caveman dictionary (Help menu) for how " + name + " works.",
        )
    return impl(interp, args, line), True

File: native/src/engine/builtins.cpp
#include "builtins.hpp"

#include <algorithm>
#include <cmath>
#include <limits>
#include <random>
#include <unordered_map>
#include <utility>

#include "errors.hpp"
#include "interpreter.hpp"
#include "textutil.hpp"
#include "words.hpp"

namespace cmc {

namespace {

using BuiltinFn = Value (*)(Interpreter&, const std::vector<Value>&, int);

std::mt19937& rng() {
    static std::mt19937 engine(std::random_device{}());
    return engine;
}

Value need_text(const Value& v, int line, const std::string& extra) {
    if (!v.is_text()) {
        throw CmcRuntimeError(
            extra + " needs text (goop), but got " + type_word(v) + ".",
            line,
            "put quotes around it, like \"hello\".");
    }
    return v;
}

Value need_pile(const Value& v, int line, const std::string& extra) {
    if (!v.is_pile()) {
        throw CmcRuntimeError(
            extra + " needs a pile (snorf), but got " + type_word(v) + ".",
            line,
            "make a pile like this: grunk xs = snorf(1, 2, 3)");
    }
    return v;
}

long long spot(const Value& v, int line, const std::string& extra) {
    return whole(v, line, extra).as_int();
}

Value bi_goop(Interpreter&, const std::vector<Value>& args, int) {
    return Value::text(show(args[0]));
}

Value bi_snorf(Interpreter&, const std::vector<Value>& args, int) {
    Value pile = Value::pile();
    pile.as_pile()->items = args;
    return pile;
}

Value bi_nom(Interpreter&, const std::vector<Value>& args, int line) {
    const Value& v = args[0];
    if (v.is_text()) return Value::integer(static_cast<long long>(utf8_length(v.as_text())));
    if (v.is_pile()) return Value::integer(static_cast<long long>(v.as_pile()->items.size()));
    throw CmcRuntimeError(
        "nom counts piles and text, but got " + type_word(v) + ".",
        line,
        "nom(pets) counts a pile. nom(\"cave\") counts letters.");
}

Value bi_skoop(Interpreter&, const std::vector<Value>& args, int line) {
    const Value& v = args[0];
    long long idx = spot(args[1], line, "skoop");
    if (!v.is_text() && !v.is_pile()) {
        throw CmcRuntimeError(
            "skoop reaches into piles and text, but got " + type_word(v) + ".",
            line,
            "make a pile like this: grunk xs = snorf(1, 2, 3)");
    }
    long long length = v.is_text() ? static_cast<long long>(utf8_length(v.as_text()))
                                   : static_cast<long long>(v.as_pile()->items.size());
    if (idx < 0) idx += length;
    if (idx < 0 || idx >= length) {
        throw CmcRuntimeError(
            "There is no spot " + show(Value::integer(idx)) + " in that "
                + (v.is_text() ? "text" : "pile") + " (it has " + std::to_string(length)
                + " spots).",
            line,
            "spots start at 0, so the last spot is " + std::to_string(length - 1) + ".");
    }
    if (v.is_text()) return Value::text(utf8_at(v.as_text(), idx));
    return v.as_pile()->items[static_cast<std::size_t>(idx)];
}

Value bi_yoink(Interpreter&, const std::vector<Value>& args, int line) {
    Value pile = need_pile(args[0], line, "yoink");
    auto& items = pile.as_pile()->items;
    if (items.empty()) {
        throw CmcRuntimeError(
            "The pile is empty, so there is nothing to yoink!",
            line,
            "check nom(pile) > 0 before yoinking.");
    }
    Value top = items.back();
    items.pop_back();
    return top;
}

Value bi_plop(Interpreter&, const std::vector<Value>& args, int line) {
    Value pile = need_pile(args[0], line, "plop");
    pile.as_pile()->items.push_back(args[1]);
    return Value::plop();
}

Value bi_shout(Interpreter&, const std::vector<Value>& args, int line) {
    return Value::text(text_upper(need_text(args[0], line, "shout").as_text()));
}

Value bi_whisper(Interpreter&, const std::vector<Value>& args, int line) {
    return Value::text(text_lower(need_text(args[0], line, "whisper").as_text()));
}

Value bi_flip(Interpreter&, const std::vector<Value>& args, int line) {
    return Value::text(utf8_reverse(need_text(args[0], line, "flip").as_text()));
}

Value bi_find(Interpreter&, const std::vector<Value>& args, int line) {
    std::string text = need_text(args[0], line, "find").as_text();
    std::string piece = need_text(args[1], line, "find").as_text();
    return Value::integer(utf8_find(text, piece));
}

Value bi_split(Interpreter&, const std::vector<Value>& args, int line) {
    std::string text = need_text(args[0], line, "split").as_text();
    std::string sep = need_text(args[1], line, "split").as_text();
    if (sep.empty()) {
        throw CmcRuntimeError(
            "split needs something to cut on, and an empty text cuts nothing.", line);
    }
    Value pile = Value::pile();
    std::size_t start = 0;
    while (true) {
        std::size_t found = text.find(sep, start);
        if (found == std::string::npos) {
            pile.as_pile()->items.push_back(Value::text(text.substr(start)));
            break;
        }
        pile.as_pile()->items.push_back(Value::text(text.substr(start, found - start)));
        start = found + sep.size();
    }
    return pile;
}

Value bi_join(Interpreter&, const std::vector<Value>& args, int line) {
    Value pile = need_pile(args[0], line, "join");
    std::string sep = args[1].is_text() ? args[1].as_text() : show(args[1]);
    std::string out;
    const auto& items = pile.as_pile()->items;
    for (std::size_t i = 0; i < items.size(); ++i) {
        if (i > 0) out += sep;
        out += show(items[i]);
    }
    return Value::text(out);
}

Value bi_what(Interpreter&, const std::vector<Value>& args, int) {
    return Value::text(type_name(args[0]));
}

Value bi_numba(Interpreter&, const std::vector<Value>& args, int line) {
    const Value& v = args[0];
    if (v.is_bool()) {
        throw CmcRuntimeError(
            "numba cannot turn gronk/nork into a number.",
            line,
            "use 1 for gronk or 0 for nork.");
    }
    if (v.is_number()) return v;
    if (v.is_text()) {
        std::string text = trim_ascii(v.as_text());
        long long as_int = 0;
        if (parse_python_int(text, as_int)) return Value::integer(as_int);
        double as_float = 0.0;
        if (parse_python_float(text, as_float)) return Value::real(as_float);
        throw CmcRuntimeError(
            "numba cannot turn \"" + v.as_text() + "\" into a number.",
            line,
            "answers that are numbers look like \"42\" or \"3.5\".");
    }
    throw CmcRuntimeError(
        "numba needs text or a number, but got " + type_word(v) + ".",
        line,
        "use it like this: numba(\"42\")");
}

Value bi_munga(Interpreter&, const std::vector<Value>& args, int line) {
    Value lo = number(args[0], line, "munga");
    Value hi = number(args[1], line, "munga");
    bool ints = lo.is_int() && hi.is_int();
    if (ints) {
        long long a = lo.as_int();
        long long b = hi.as_int();
        if (a > b) std::swap(a, b);
        std::uniform_int_distribution<long long> dist(a, b);
        return Value::integer(dist(rng()));
    }
    double a = lo.as_number();
    double b = hi.as_number();
    if (a > b) std::swap(a, b);
    std::uniform_real_distribution<double> dist(a, b);
    return Value::real(dist(rng()));
}

Value bi_round(Interpreter&, const std::vector<Value>& args, int line) {
    Value v = number(args[0], line, "round");
    double d = v.as_number();
    if (d >= 0) return Value::integer(static_cast<long long>(std::floor(d + 0.5)));
    return Value::integer(-static_cast<long long>(std::floor(-d + 0.5)));
}

Value bi_flat(Interpreter&, const std::vector<Value>& args, int line) {
    return Value::integer(static_cast<long long>(std::floor(number(args[0], line, "flat").as_number())));
}

Value bi_roof(Interpreter&, const std::vector<Value>& args, int line) {
    return Value::integer(static_cast<long long>(std::ceil(number(args[0], line, "roof").as_number())));
}

Value bi_abs(Interpreter&, const std::vector<Value>& args, int line) {
    Value v = number(args[0], line, "abs");
    if (v.is_int()) {
        long long n = v.as_int();
        if (n == std::numeric_limits<long long>::min()) return Value::real(std::fabs(v.as_number()));
        return Value::integer(n < 0 ? -n : n);
    }
    return Value::real(std::fabs(v.as_float()));
}

Value bi_small(Interpreter&, const std::vector<Value>& args, int line) {
    Value a = number(args[0], line, "small");
    Value b = number(args[1], line, "small");
    if (a.is_int() && b.is_int()) return a.as_int() <= b.as_int() ? a : b;
    return a.as_number() <= b.as_number() ? a : b;
}

Value bi_big(Interpreter&, const std::vector<Value>& args, int line) {
    Value a = number(args[0], line, "big");
    Value b = number(args[1], line, "big");
    if (a.is_int() && b.is_int()) return a.as_int() >= b.as_int() ? a : b;
    return a.as_number() >= b.as_number() ? a : b;
}

Value bi_root(Interpreter&, const std::vector<Value>& args, int line) {
    Value v = number(args[0], line, "root");
    if (v.as_number() < 0) {
        throw CmcRuntimeError(
            "root cannot work on negative numbers (nothing times itself makes " + show(v) + ").",
            line,
            "use abs(value) first if you want the size without the minus.");
    }
    double result = std::sqrt(v.as_number());
    if (std::floor(result) == result) return Value::integer(static_cast<long long>(result));
    return Value::real(result);
}

Value bi_pow(Interpreter&, const std::vector<Value>& args, int line) {
    Value base = number(args[0], line, "pow");
    Value times = number(args[1], line, "pow");
    double result = std::pow(base.as_number(), times.as_number());
    if (std::isfinite(result) && std::floor(result) == result
        && std::fabs(result) < 9.0e18) {
        return Value::integer(static_cast<long long>(result));
    }
    return Value::real(result);
}

Value bi_wait(Interpreter& interp, const std::vector<Value>& args, int line) {
    Value ms = number(args[0], line, "wait");
    double value = ms.as_number();
    if (value < 0) value = 0;
    interp.wait_ms(value);
    return Value::plop();
}

const std::unordered_map<std::string, BuiltinFn>& impls() {
    static const std::unordered_map<std::string, BuiltinFn> table = {
        {"goop", bi_goop},       {"snorf", bi_snorf},   {"nom", bi_nom},
        {"skoop", bi_skoop},     {"yoink", bi_yoink},   {"plop", bi_plop},
        {"shout", bi_shout},     {"whisper", bi_whisper}, {"flip", bi_flip},
        {"find", bi_find},       {"split", bi_split},   {"join", bi_join},
        {"what", bi_what},       {"numba", bi_numba},   {"munga", bi_munga},
        {"round", bi_round},     {"flat", bi_flat},     {"roof", bi_roof},
        {"abs", bi_abs},         {"small", bi_small},   {"big", bi_big},
        {"root", bi_root},       {"pow", bi_pow},       {"wait", bi_wait},
    };
    return table;
}

}

bool is_builtin(const std::string& name) {
    return impls().count(name) > 0;
}

Value call_builtin(Interpreter& interp, const std::string& name, const std::vector<Value>& args,
                   int line) {
    auto it = impls().find(name);
    if (it == impls().end()) return Value::plop();
    auto arity_it = words::arity().find(name);
    int low = 0;
    int high = 0;
    if (arity_it != words::arity().end()) {
        low = arity_it->second.first;
        high = arity_it->second.second;
    }
    if (static_cast<int>(args.size()) < low || static_cast<int>(args.size()) > high) {
        std::string wanted = (low == high) ? std::to_string(low)
                                           : std::to_string(low) + " to " + std::to_string(high);
        std::string plural = (wanted != "1") ? "s" : "";
        throw CmcRuntimeError(
            "'" + name + "' wants " + wanted + " thing" + plural + ", but got "
                + std::to_string(args.size()) + ".",
            line,
            "look at the caveman dictionary (Help menu) for how " + name + " works.");
    }
    return it->second(interp, args, line);
}

}
